Generated C code issue with variant objects defined in blocks.

The C backend gets confused on what object you mean if type definition is inside a block. Maybe its looking at wrong type with the same name?

### Example
```nim
type
  MyObjectKind = enum
    SomeKind

block:
  type
    MyObject = ref object
      case kind: MyObjectKind:
        of SomeKind:
          c: string

  var obj = MyObject(kind:SomeKind)
  obj.c = "some string"

block:
  type
    MyObject = ref object
      kind: MyObjectKind
      c: string

  var obj = MyObject(kind:SomeKind)
  obj.c = "some string"
```

### Current Output
```
/root/.cache/nim/uuu2_r/@muuu2.nim.c: In function ‘NimMainModule’:
/root/.cache/nim/uuu2_r/@muuu2.nim.c:377:41: error: ‘tyObject_MyObjectcolonObjectType___i0gd8Xwpb5mzDLo8lUmaLA {aka struct tyObject_MyObjectcolonObjectType___i0gd8Xwpb5mzDLo8lUmaLA}’ has no member named ‘c’
   T6_ = (*obj__IgBTMVGOV1i6k1DxUIPA6A_2).c; (*obj__IgBTMVGOV1i6k1DxUIPA6A_2).c = copyStringRC1(((NimStringDesc*) &TM__Uvt1GVKLB2r8cLJi4I9bK4A_7));
                                         ^
/root/.cache/nim/uuu2_r/@muuu2.nim.c:377:77: error: ‘tyObject_MyObjectcolonObjectType___i0gd8Xwpb5mzDLo8lUmaLA {aka struct tyObject_MyObjectcolonObjectType___i0gd8Xwpb5mzDLo8lUmaLA}’ has no member named ‘c’
   T6_ = (*obj__IgBTMVGOV1i6k1DxUIPA6A_2).c; (*obj__IgBTMVGOV1i6k1DxUIPA6A_2).c = copyStringRC1(((NimStringDesc*) &TM__Uvt1GVKLB2r8cLJi4I9bK4A_7));
                                                                             ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -O3 -fno-strict-aliasing -fno-ident   -I'/root/.choosenim/toolchains/nim-#devel/lib' -I/p/try -o /root/.cache/nim/uuu2_r/@muuu2.nim.c.o /root/.cache/nim/uuu2_r/@muuu2.nim.c' failed with exit code: 1
```

### Expected Output

It just compiles.

### Solutions

Just naming the types differently works.

```nim
type
  MyObjectKind = enum
    SomeKind

block:
  type
    MyObject = ref object
      case kind: MyObjectKind:
        of SomeKind:
          c: string

  var obj = MyObject(kind:SomeKind)
  obj.c = "some string"

block:
  type
    MyObject2 = ref object
      kind: MyObjectKind
      c: string

  var obj = MyObject2(kind:SomeKind)
  obj.c = "some string"
```

```
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-01-27
Copyright (c) 2006-2021 by Andreas Rumpf

active boot switches: -d:release
```

I've been able to reproduce this both on 1.5.1, and 1.4.4. If you use any kind of object, it also seems to crash, it doesn't need to be a variant.
My error looks like 
```
/home/ajusa/.cache/nim/intro_d/@m..@s..@s.nimble@spkgs@sprint-0.2.0@sprint.nim.c: In function ‘prettyPrint__VZa9a9cnwLdtQNvQrlfUad3g_2’:
/home/ajusa/.cache/nim/intro_d/@m..@s..@s.nimble@spkgs@sprint-0.2.0@sprint.nim.c:1850:61: error: ‘tyObject_HttpcolonObjectType___kJvy9czVI2Oyw2NP5bv5k7Q’ has no member named ‘headers’
 1850 |  eX60gensym25__4 = prettyPrint__9aqV4s3O1K50PdR07Vh1KYA((*x).headers, ((NI) 0), NIM_FALSE);
      |                                                             ^
/home/ajusa/.cache/nim/intro_d/@m..@s..@s.nimble@spkgs@sprint-0.2.0@sprint.nim.c:1876:52: error: ‘tyObject_HttpcolonObjectType___kJvy9czVI2Oyw2NP5bv5k7Q’ has no member named ‘headers’
 1876 |   T107_ = prettyPrint__9aqV4s3O1K50PdR07Vh1KYA((*x).headers, (NI)(TM__Impvaj6MClV2Sg0bUNMELQ_101), NIM_TRUE);
      |                                                    ^
CC: ../../.nimble/pkgs/mustache-0.3.2/mustachepkg/render.nim
CC: ../../.nimble/pkgs/nimib-#20a29751e73ffd4a9d60c45157ea3857d91bd5ba/nimib/types.nim
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/ajusa/.choosenim/toolchains/nim-1.4.4/lib -I/home/ajusa/Documents/binarylang-fun -o /home/ajusa/.cache/nim/intro_d/@m..@s..@s.nimble@spkgs@sprint-0.2.0@sprint.nim.c.o /home/ajusa/.cache/nim/intro_d/@m..@s..@s.nimble@spkgs@sprint-0.2.0@sprint.nim.c' failed with exit code: 1
```

Same issue with generic types:
```nim

import macros

block:
  type
    MyBaseObj[T] = object of RootObj
      test: seq[T]
    MyObj = object
      attr: MyBaseObj[string]

  var myObj: MyObj

block:
  type
    MyBaseAttrObj[T] = object of RootObj
      test: seq[T]
    MyBaseObj[T] = object of RootObj
      myBaseObj: MyBaseAttrObj[T]
    MyObj = object
      attr: MyBaseObj[string]

  var myObj: MyObj
```

Fails with:
```bash
/home/dankrad/.cache/nim/blockissue2_d/@mblockissue2.nim.c: In function ‘TM__JBJ7iAN5wQWK3NIW4CNX9aQ_3’:
/home/dankrad/.cache/nim/blockissue2_d/@mblockissue2.nim.c:152:65: error: ‘tyObject_MyBaseObj__MpV250MmznRzvPQBxwRh5A’ has no member named ‘myBaseObj’
  152 |         nimGCvisit((void*)myObj__k6Sw9aSdp5id9ajw1VQM3kEw_2.attr.myBaseObj.test, 0);
      |                                                                 ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/opt/Nim/lib -I/home/dankrad/Desktop -o /home/dankrad/.cache/nim/blockissue2_d/@mblockissue2.nim.c.o /home/dankrad/.cache/nim/blockissue2_d/@mblockissue2.nim.c' failed with exit code: 1
```

--- a/tests/ccgbugs/tccgissues.nim
+++ b/tests/ccgbugs/tccgissues.nim
@@ -12,3 +12,46 @@ proc varargsToSeq(vals: varargs[int32]): seq[int32] =
     result.add v
 
 echo varargsToSeq(1, 2, 3, 4)
+
+type # bug #16845
+  MyObjectKind = enum
+    SomeKind
+
+block:
+  type
+    MyObject = ref object
+      case kind: MyObjectKind:
+        of SomeKind:
+          c: string
+
+  var obj = MyObject(kind:SomeKind)
+  obj.c = "some string"
+
+block:
+  type
+    MyObject = ref object
+      kind: MyObjectKind
+      c: string
+
+  var obj = MyObject(kind:SomeKind)
+  obj.c = "some string"
+
+block:
+  type
+    MyBaseObj[T] = object of RootObj
+      test: seq[T]
+    MyObj = object
+      attr: MyBaseObj[string]
+
+  var myObj: MyObj
+
+block:
+  type
+    MyBaseAttrObj[T] = object of RootObj
+      test: seq[T]
+    MyBaseObj[T] = object of RootObj
+      myBaseObj: MyBaseAttrObj[T]
+    MyObj = object
+      attr: MyBaseObj[string]
+
+  var myObj: MyObj

